src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+namespace GraphQL.Subscription;
+
+internal static class ObservableExtensions
+{
+    /// <summary>
+    /// <para>
+    /// Applies an asynchronous transformation on data events from an <see cref="IObservable{T}"/>.
+    /// Maintains the order of the events produced by the <see cref="IObservable{T}"/>
+    /// whether they are data, error or completion notifications.
+    /// </para>
+    /// <para>
+    /// Ensures that after an <see cref="IObserver{T}"/> stream has been disposed,
+    /// no more events will be raised (data, error or completion), and signals
+    /// pending asynchronous transformations that a cancellation has been requested.
+    /// </para>
+    /// <para>
+    /// Exceptions passed by the source through <see cref="IObserver{T}.OnError(Exception)"/> or
+    /// generated by <paramref name="transformNext"/> are handled by <paramref name="transformError"/>.
+    /// </para>
+    /// </summary>
+    public static IObservable<TOut> SelectCatchAsync<TIn, TOut>(this IObservable<TIn> observable, Func<TIn, CancellationToken, Task<TOut>> transformNext, Func<Exception, CancellationToken, Task<Exception>> transformError)
+        => new SelectCatchAsyncObservable<TIn, TOut>(observable, transformNext, transformError);
+
+    private class SelectCatchAsyncObservable<TIn, TOut> : IObservable<TOut>
```

Comment:
Do I understand correctly that `SelectCatchAsyncObservable` exists only because we need **async** projection from object to ExecutionResult? In case of sync projection we could use `Select` Rx LINQ operator (of course copy-pasted here).

Reply:
Basically yes.  Sync projection is easy.  But we need to maintain the order of the completed async tasks "on the way out" in the order they originally arrived "on the way in".

There was some custom code unique to GraphQL.NET with how exceptions were dealt with, but that is basically gone now.

I'm not positive, but I think basically it just encompasses this functionality from System.Reactive:

```cs
public static IObservable<TOut> SelectCatchAsync<TIn, TOut>(
    this IObservable<TIn> observable,
    Func<TIn, CancellationToken, Task<TOut>> transformNext,
    Func<Exception, CancellationToken, Task<Exception>> transformError)
{
    return observable.SelectAsync(transformNext).CatchAsync(error => throw await transformError(error));
}
```

Adding ObservableExtensions.cs only serves to eliminate System.Reactive.  It has nothing to do with the bug fixes listed in #3002 or vice versa.

Reply:
If I understand correctly `SelectAsync/CatchAsync` above is a pseudo code. I can not see such methods in Rx. Also `Catch` method conceptually does different things
```
 /// <summary>
        /// Continues an observable sequence that is terminated by an exception with the next observable sequence.
        /// </summary>
        /// <typeparam name="TSource">The type of the elements in the source sequence and handler sequence.</typeparam>
        /// <param name="first">First observable sequence whose exception (if any) is caught.</param>
        /// <param name="second">Second observable sequence used to produce results when an error occurred in the first sequence.</param>
        /// <returns>An observable sequence containing the first sequence's elements, followed by the elements of the second sequence in case an exception occurred.</returns>
        /// <exception cref="ArgumentNullException"><paramref name="first"/> or <paramref name="second"/> is null.</exception>
        public static IObservable<TSource> Catch<TSource>(this IObservable<TSource> first, IObservable<TSource> second)
``` 

Does not matter.


Regarding this comment:
> Adding ObservableExtensions.cs only serves to eliminate System.Reactive.

I would like you to put such a comment on that file. This idea can also be reflected in the documentation, noting that we get rid of very heavy dependency.

Reply:
I added the note to the file.